### PR TITLE
Add Reference ILU(0)

### PR DESCRIPTION
**Pull request by upsj**
This PR adds a reference ILU(0) and IC(0) implementation with accompanying tests. The tests are exactly the ones we use for ParIlu/ParIc.

### CODE REVIEW

**Review by Slaedr — COMMENTED** [review]
LGTM!

**Review by Slaedr — APPROVED** [review]
And once more for good luck, LGTM! 😅

File: reference/test/factorization/ic_kernels.cpp [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+
+    std::shared_ptr<const gko::ReferenceExecutor> ref;
+    std::shared_ptr<const gko::Executor> exec;
+    std::shared_ptr<Csr> identity;
+    std::shared_ptr<Csr> banded;
+    std::shared_ptr<Csr> banded_l_expect;
+    std::shared_ptr<Csr> mtx_system;
+    std::unique_ptr<Csr> mtx_l_system;
+    std::unique_ptr<Coo> mtx_l_system_coo;
+    std::unique_ptr<Csr> mtx_l_init_expect;
```
Comment (MarcelKoch) [inline reference/test/factorization/ic_kernels.cpp:130]: This variable is not used, is there a test missing or can this be removed?

File: reference/test/factorization/ilu_kernels.cpp [hunk 2]
```diff
@@ -0,0 +1,505 @@
...
+        mtx_small->convert_to(gko::lend(tmp_csr));
+        mtx_csr_small = std::move(tmp_csr);
+        auto tmp_csr2 = Csr::create(exec);
+        mtx_small2->convert_to(gko::lend(tmp_csr2));
+        mtx_csr_small2 = std::move(tmp_csr2);
+    }
+
+    std::shared_ptr<const gko::ReferenceExecutor> ref;
+    std::shared_ptr<const gko::Executor> exec;
+    std::shared_ptr<const Csr> empty_csr;
```
Comment (MarcelKoch) [inline reference/test/factorization/ilu_kernels.cpp:189]: This is also not used elsewhere, perhaps a missing test?

File: reference/test/factorization/ilu_kernels.cpp [hunk 3]
```diff
@@ -0,0 +1,505 @@
...
+
+    GKO_ASSERT_MTX_NEAR(l_factor, this->big_l_expected, r<value_type>::value);
+    GKO_ASSERT_MTX_NEAR(u_factor, this->big_u_expected, r<value_type>::value);
+}
+
+
+TYPED_TEST(Ilu, GenerateForDenseBigSort)
+{
+    using value_type = typename TestFixture::value_type;
+    auto factors = this->ilu_factory_skip->generate(this->mtx_big);
```
Comment (MarcelKoch) [inline reference/test/factorization/ilu_kernels.cpp:439]:
```suggestion
    auto factors = this->ilu_factory_sort->generate(this->mtx_big);
```

**Review by MarcelKoch — APPROVED** [review]
lgtm, I've left some comments on the tests, which could be addressed

File: reference/test/factorization/ic_kernels.cpp (hunk 1 above)
Comment (upsj) [reply reference/test/factorization/ic_kernels.cpp:130]: Good catch, those are left-over from ParIC tests.

**Review by thoasm — APPROVED** [review]
LGTM, I only have minor nits.

File: reference/test/factorization/ilu_kernels.cpp (hunk 3 above)
Comment (upsj) [reply reference/test/factorization/ilu_kernels.cpp:439]: Great catch, that was also an issue in the ParILU tests

File: reference/factorization/ic_kernels.cpp [hunk 4]
```diff
@@ -46,7 +52,52 @@ namespace ic_factorization {
 
 template <typename ValueType, typename IndexType>
 void compute(std::shared_ptr<const DefaultExecutor> exec,
-             matrix::Csr<ValueType, IndexType> *m) GKO_NOT_IMPLEMENTED;
+             matrix::Csr<ValueType, IndexType> *m)
+{
+    vector<IndexType> diagonals{m->get_size()[0], -1, {exec}};
```
[the remark this answers is not in the document]
Comment (upsj) [reply reference/factorization/ic_kernels.cpp:57]: I wanted to initialize the data to be able to check that we reached all diagonals in Debug. I thought we always used this pattern for the executor allocator, but it seems I was wrong, so I'll change it.

**Pull request merged**